### PR TITLE
localstack doesn't like S3 user metadata in snake_case

### PR DESCRIPTION
**Pull request by akash1810**
## What does this change?
In https://github.com/guardian/grid/pull/2927 we're enabling running Grid locally in [localstack](https://github.com/localstack/localstack). It's quite a big PR which makes it hard to properly reason about. This change can be shipped independently. Small and often over big bang.

Localstack doesn't support snake_case in S3 object metadata, so move to kebab-case:
- Write metadata in kebab-case
- Read metadata as kebab-case and fallback to snake_case for backwards compatibility

More info from localstack [here](https://github.com/localstack/localstack/issues/459).

## How can success be measured?
no-op

## Screenshots (if applicable)
n/a

## Who should look at this?

@guardian/digital-cms

## Tested?
- [x] locally
- [x] on TEST

### CODE REVIEW

File: cropper/app/lib/CropStore.scala [hunk 1]
```diff
@@ -56,15 +58,19 @@ class CropStore(config: CropperConfig) extends S3ImageStorage(config) {
...
             source <- userMetadata.get("source")
-            x      <- userMetadata.get("bounds_x").map(_.toInt)
-            y      <- userMetadata.get("bounds_y").map(_.toInt)
-            w      <- userMetadata.get("bounds_w").map(_.toInt)
-            h      <- userMetadata.get("bounds_h").map(_.toInt)
+
+            // we've moved to kebab-case as localstack doesn't like `_`
+            // fallback to reading old values for older crops
+            // see https://github.com/localstack/localstack/issues/459
+            x      <- Try(userMetadata.getOrElse("bounds-x", userMetadata("bounds_x"))).toOption.map(_.toInt)
```
Comment (akash1810) [inline cropper/app/lib/CropStore.scala:65]: Wrap in a `Try` as the `getOrElse` fallback of `userMetadata("bounds_x")` isn't guaranteed to succeed.

Comment (AWare) [reply cropper/app/lib/CropStore.scala:65]: I'd like to see this wrapped in a function.

Comment (akash1810) [reply cropper/app/lib/CropStore.scala:65]: I did this but it made it less readable and was less obvious what was going on imo:

```scala
private def getMetadataValue(map: Map[String, String], preferredKey: String, fallbackKey: String): Option[String] = {
  Try(map.getOrElse(preferredKey, map(fallbackKey))).toOption
}

val updatedCrop = for {
  x <- getMetadataValue(userMetadata, "bounds-x", "bounds_x").map(_.toInt)
  y <- getMetadataValue(userMetadata, "bounds-y", "bounds_y").map(_.toInt)
  ... etc
}
```

Comment (akash1810) [reply cropper/app/lib/CropStore.scala]: I think this is down to naming; `getOrElseOrNone` makes it clearer.

File: cropper/app/lib/CropStore.scala [hunk 2]
```diff
@@ -44,6 +46,11 @@ class CropStore(config: CropperConfig) extends S3ImageStorage(config) {
     }
   }
 
+  private def getOrElseOrNone(theMap: Map[String, String], preferredKey: String, fallbackKey: String): Option[String] = Try(
+    // Return the `preferredKey` value in `theMap` or the `fallbackKey` or `None`
+    theMap.getOrElse(preferredKey, theMap(fallbackKey))
+  ).toOption
```
Comment (sihil) [inline cropper/app/lib/CropStore.scala:52]: Using Try is a weird pattern here. In essence it is throwing and catching and processing an exception which seems to be overkill and confusing. This seems a lot clearer to me.

```suggestion
  private def getOrElseOrNone(theMap: Map[String, String], preferredKey: String, fallbackKey: String): Option[String] = {
    // Return the `preferredKey` value in `theMap` or the `fallbackKey` or `None`
    theMap.get(preferredKey).orElse(theMap.get(fallbackKey))
  }
```

Comment (sihil) [reply cropper/app/lib/CropStore.scala:52]: It seems barely longer than the call to this function so it's tempting to put this inline.

File: cropper/app/lib/CropStore.scala [hunk 3]
```diff
@@ -56,15 +63,19 @@ class CropStore(config: CropperConfig) extends S3ImageStorage(config) {
...
             // Note: if any is missing, the entry won't be registered
             source <- userMetadata.get("source")
-            x      <- userMetadata.get("bounds_x").map(_.toInt)
-            y      <- userMetadata.get("bounds_y").map(_.toInt)
-            w      <- userMetadata.get("bounds_w").map(_.toInt)
-            h      <- userMetadata.get("bounds_h").map(_.toInt)
+
+            // we've moved to kebab-case as localstack doesn't like `_`
+            // fallback to reading old values for older crops
+            // see https://github.com/localstack/localstack/issues/459
```
Comment (sihil) [inline cropper/app/lib/CropStore.scala:69]: They claim this is 'kludged'. Possibly not merged. Makes me sad that we are working around an issue in a localstack but 🤷

**Pull request merged**